### PR TITLE
Work Toward a more Modular Chart Library

### PR DESCRIPTION
**Pull request by breck7**
This change is a step toward modularizing the Grapher client charting library as first proposed here (https://github.com/owid/owid-grapher/issues/84) to help improve developer productivity as well as potentially enable new use cases like this (https://github.com/owid/owid-grapher/issues/362).

This change does not change any existing behavior but adds:

- An easy way to get the config code for a chart from the admin panel (aka "Freezing").
Screen Shot 2020-03-31 at 12 49 52 PM

- A simple HTML "sandbox" page for loading just the chart library and a chart (aka "Thawing")
Screen Shot 2020-03-31 at 12 53 06 PM

- The option to use a static URL as the data source for a chart. The default behavior is to generate the URL from Variable IDs but if an "externalDataUrl" is defined that will be used instead. This might have other benefits in the future but for now at least allows you to debug a prod chart locally without syncing the DB.

- (very limited) ability to create a Grapher chart using Tabular data. When the `tabularData` param is provided a method will convert the provided data to our Variable format. Switching to a tabular model would be a big change so I thought we'd take small steps so we can do it without breaking anything and evaluate the benefits as we go. I think a tabular data model has a few benefits. I think it becomes simpler to debug/inspect all stages of the visualization pipeline. I think it will make cleaning up the year/day code easier. Also, it will align with the way many other data tools including Vega and Tidyverse operate so we might be able to interoperate better with other tools and organizations in the data-vis ecosystem.
Screen Shot 2020-03-31 at 12 54 43 PM

This change also removes the "VariableData" class without sacrificing any functionality.

### CODE REVIEW

File: public/sandbox/index.html [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        <link rel="stylesheet" href="http://localhost:8090/css/commons.css" />
+        <link rel="stylesheet" href="http://localhost:8090/css/owid.css" />
+    </head>
+    <body class="ChartPage">
+        <main>
+            <figure data-grapher-src=""></figure>
+        </main>
+        <script src="http://localhost:8090/js/commons.js"></script>
+        <script src="http://localhost:8090/js/owid.js"></script>
+        <script src="chart.js"></script>
```
Comment (danielgavrilov) [inline public/sandbox/index.html:19]:
```suggestion
        <script src="chart.sample.js"></script>
```

File: charts/ChartConfig.tsx [hunk 2]
```diff
@@ -215,6 +243,60 @@ export class ChartConfig {
...
     @observable.ref useTimelineDomains = false
 
+    @observable.ref variablesById: { [id: number]: Variable } = {}
+    @observable.ref entityMetaById: { [id: number]: EntityMeta } = {}
+
+    @computed get availableEntities(): string[] {
+        return keys(this.entityMetaByKey)
+    }
+
+    @action.bound private async downloadData() {
```
[the remark this answers is not in the document]
Comment (breck7) [reply charts/ChartConfig.tsx:253]: Yes! I was looking for a TS refactoring tool that could detect all potential privates project wide and add them but didn't anything yet, so for now just adding them as I go.

File: public/sandbox/index.html (hunk 1 above)
Comment (breck7) [reply public/sandbox/index.html:19]: Ah I should document this. This one is sort of like ".env.example" in that I expect you to create a chart.js yourself (from a freeze).

**Comment by MarcelGerber** [comment]
@breck7 did you intend for this to already land in master? Looks like some commits are already in there, e.g. https://github.com/owid/owid-grapher/commit/ff64befbb73d9527650e9c32c5d90ca792e04a5e

**Comment by danielgavrilov** [comment]
Yeah I will close this PR since it's already merged, reopen if I got it wrong.

**Comment by breck7** [comment]
Yes @MarcelGerber I merged it in locally and pushed to master because the GitHub merge was failing CI and it just seemed easier to merge the conflicts locally. Thanks for closing @danielgavrilov.

**Pull request closed without merging**